Add support for L3-dependent L4 policies.

joestringer opened the pull request:
When specifying policies for ingress, users may provide `FromEndpoints` or `ToPorts`. These apply constraints on the access between endpoints based upon source labels (at L3) or destination ports (at L4). Up until this point, if a single rule specified both `FromEndpoints` and `ToPorts`, the policy would allow the union of these constraints rather than the intersection. For example, if you specified a single rule that required `FromEndpoints:k8s:id=frontend` as well as `ToPorts:80`, then any traffic from `k8s:id=frontend` would be allowed regardless of ports, and any traffic destined to port 80 would be allowed regardless of the source labels.

This PR fixes the behaviour to now apply the intersection: traffic must match all of the FromEndpoints constraints *AND* the ToPorts constraints to be allowed. Within Cilium, this means updating the Label-based policy lookup to return 'api.Undecided' when it finds a rule that specifies L4 constraints; updating the L4 lookup to also be smart enough to apply Label constraints; and ensuring that these policies can be correctly reflected down to the BPF maps. This last part is performed by extending the BPF PolicyMap to also perform lookups using L4 information. Cilium will expand out L4 policies into {SecurityID,port,protocol} tuples into this map.

In BPF, the PolicyMap may now hold entries which specify Ports and Protocols in addition to the SourceIdentities that have been used so far. The datapath logic first uses the source identity plus the packet's L4 information (port,protocol) to perform a lookup. If this matches, because for instance there is an L4-only policy applied and the daemon has pushed the policy down, or if it matches an L4-dependent policy which was pushed down, then traffic will be allowed. If this initial lookup fails, then the datapath will perform a second lookup in the PolicyMap, this time without L4 information. This is the regular L3/Label-based lookup path that already existed. Note that the new L4 or L3+L4 combined lookup only occurs during connection tracking entry creation, so the extra lookup cost should be bound to the first packet of each connection.

Fixes: #789
Fixes: #1064
Fixes: #1217 
Fixes: #1496 

Future work:
- Revisit whether the BPF L4 port policy code generation could be removed (it's still responsible for Egress ToPort policy, though)
- Add CIDR-dependent L4 policy

tgraf committed:
k8s: Reject policy rules with combined L3/L4 policies

- Reject k8s NetworkPolicy rules with combined L3/L4 policy rules.

- Don't generate api.Rule with with combined L3/L4 policy rules.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Add optional L4 destination port and protocol to policy map

If HAVE_L4_POLICY is set, the policy map will be looked up with the
destination port and protocol set. Allows to establish L4 policies
which are tied to a particular source identity.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Co-authored-by: Joe Stringer <joe@covalent.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

amreshakim committed:
l4: Add L3 dependency to L4 policies

This commit extends the L4Filter to include a list of EndpointSelectors,
which enables a future commit to narrow down which endpoints the L4
policy will ultimately apply to. For now, the existing paths will follow
the paths with the WildcardEndpointSelector, which allows an L4Filter to
apply to all traffic (as the L4-only policies already do today).

Signed-off-by: Amre Shakimov <amre@covalent.io>
Co-authored-by: Joe Stringer <joe@covalent.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

amreshakim committed:
add github.com/mitchellh/hashstructure to vendor

Signed-off-by: Amre Shakimov <amre@covalent.io>

joestringer committed:
k8s: Refactor parsing NetworkPolicyPeer

No functional change.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Refactor AllowsRLocked to resolve L4 policy

When the PolicyRepository's AllowsRLocked() function is called, based on
the naming convention and documentation we'd expect to determine that,
for the entire SearchContext provided, we would receive a complete
Label/L3/L4 decision based on the full policy. However, so far it has
only represented the Label portion of the policy. The EndPoint package
uses this to populate the BPF maps for label access, so split the label
portion into a separate function and reuse it there. Then, make
AllowsRLocked() great again by expanding its scope to include L4 policy
resolution.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: Sanitize protocol capitalization

Always set the protocol to lower case when interpreting it from a k8s
policy.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
k8s: Extend network policy tests for rules union

Add an extra test to ensure that the example 4 (union of two rules)
policy is being correctly applied as a union, rather than intersection.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Test L4Filter creation builds L7filters

This simple test just checks that when specifying L7 rules during
L4Filter creation, regardless of whether it's ingress or egress, and
whether there are endpoints specified or not, the L4Filter ends up with
an L7 rule.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf committed:
policy: Reject policy rules with combined L3/L4 policies

Reserve the right to support L4/L7 policies which depend on L3 matching
in the future when the datapath supports this. For now, reject any
such rules but document that this will be supported in the future.

Signed-off-by: Thomas Graf <thomas@cilium.io>
Co-authored-by: Joe Stringer <joe@covalent.io>
Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
policy: Add support for L3-dependent L4 policy

The existing L3 (label-based) policy resolution functions such as
AllowsLabelAccess(ctx) and CanReachRLocked(ctx) will now check the
target rules for whether the rule also applies L4 constraints, and in
these cases return an api.Undecided value.

The L4PolicyMap now includes both regular L4-only L4Filters, which do
not apply any constraints on the source endpoints, as well as the new
L3+L4 combined filters. The existing L4-only test functions will return
failure if they find an L4Filter that constrains which endpoints must
originate the traffic.

Repository.AllowsRLocked() now performs first, a label-based lookup
which may permit traffic regardless of L4 rules. Secondly, it runs the
L4Egress and L4Ingress policy lookups, which for Egress perfoms an
L4-only lookup, whereas for Ingress performs the L3+L4 combined lookup,
which may also explicitly Allow or Deny traffic. If either of these L4
policies deny the traffic, the traffic is denied. Otherwise, if at least
one of the policies explicitly allows the traffic, then the traffic will
overall be allowed.

Signed-off-by: Joe Stringer <joe@covalent.io>

tgraf commented on pkg/policy/api/rule.go:
```diff
@@ -73,6 +73,13 @@ type Rule struct {
 //   the effects of any Requires field in any rule will apply to all other
 //   rules as well.
+//
+// - For now, combining ToPorts, FromCIDR, and FromEndpoints in the same rule
```
Can we just drop this commit? I think you will undo this anyway in a later commit.

tgraf commented on Documentation/gettingstarted.rst:
```diff
@@ -220,7 +220,7 @@ We can achieve that with the following Kubernetes NetworkPolicy:
         - podSelector:
             matchLabels:
               id: app2
-        ports:
```
Are you undoing all of these later on again? I did this commit to stop showing L3/L4 combined policies in the guides when we don't support them. We never merged it though. Given that we now support it, this commit is no longer required. So we can either just skip it here and rebase or undo it at the end.

joestringer replied:
Let me look at skipping it here and see whether anything falls out.

tgraf reviewed (dismissed):
This is amazing work. I spotted no major flaws. The main question is on the intermediate commits which change tests and docs.

joestringer commented on bpf/lib/policy.h:
```diff
@@ -24,14 +24,37 @@
+		__sync_fetch_and_add(&policy->packets, 1);
+		__sync_fetch_and_add(&policy->bytes, skb->len);
+		return TC_ACT_OK;
+	}
```
Should we consider tracing either the positive / negative match cases for L4 policy?

tgraf replied:
I think what you have is fine to get started.

joestringer commented on pkg/k8s/network_policy.go:
```diff
@@ -118,7 +122,21 @@ func ParseNetworkPolicy(np *networkingv1.NetworkPolicy) (api.Rules, error) {
+			//   sources (traffic not restricted by source).
+			//
+			// FIXME-L3-L4: Once supported, all rule should be combined into
+			// l4 policies
```
This comment can be dropped.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -134,6 +167,22 @@ func (l4 L4Filter) String() string {
 	return string(b)
 }
 
+func (l4 L4Filter) matchesLabels(labels labels.LabelArray) bool {
```
@tgraf I wanted to double-check, if we have a policy like:

```
  [{                                                                            
    "endpointSelector": {"matchLabels":{"id.bar":""}},                          
    "ingress": [{                                                               
      "fromEndpoints": [                                                        
          {"matchLabels":{"reserved:host":""}},                                     
          {"matchLabels":{"id.foo":""}
      }]                                                                           
    }]                                                                    
  }]
```

Then this means that we should apply the policy at `id.bar` that the fromEndpoint must include either the label `reserved:host` or `id.foo`. It does not require both labels. Correct?

tgraf replied:
Correct, fromEndpoints is a slice and each member is an allowed combination of labels. `matchLabels` can include multiple labels if more than label must match.

joestringer commented on pkg/policy/rule.go:
```diff
@@ -183,11 +232,15 @@ func (r *rule) canReach(ctx *SearchContext, state *traceState) api.Decision {
+					ctx.PolicyTrace("+       No L4 restrictions; allowing\n")
+					return api.Allowed
+				}
+				ctx.PolicyTrace("        Restricted to specific L4 destinations; skip during Label stage\n")
```
Any preferences on how this trace log should indicate that the label stage cannot select this rule because it includes L4 requirements? In practice I think this will end up with the trace output stating later on that "Label verdict: undecided", but I don't know if the above line clearly describes why this decision can't be made yet.

Related: #1610

ianvernon replied:
maybe specify how you cannot make a purely L3 / label-specific decision whether policy allows or denies? E.g.:
```
L3 traffic restricted to specific L4 destinations; cannot determine whether policy allows traffic purely based off of L3
```

joestringer replied:
How about this, a little more terse: 
```Rule restricts traffic to specific L4 destinations; deferring policy decision to L4 policy stage```

joestringer reviewed (commented):
I posted a few questions that I have for reviewers of this patchset.

joestringer committed:
cmd: Don't require -d to specify --dport

Previously, to perform L4 policy tracing through the Cilium policy, the
user had to specify the "-d" flag. However, if for instance someone just
wants to perform L4 policy tracing purely using k8s tags, this doesn't
make sense. Loosen this requirement.

Signed-off-by: Joe Stringer <joe@covalent.io>

aanm commented on pkg/k8s/network_policy.go:
```diff
@@ -84,19 +84,16 @@ func parseNetworkPolicyPeer(namespace string, peer *networkingv1.NetworkPolicyPe
-			ingress.FromEndpoints = append(ingress.FromEndpoints, all)
-		} else {
+		ingress := api.IngressRule{}
+		if iRule.From != nil || len(iRule.From) > 0 {
```
Why isn't it this logic: ` iRule.From != nil && len(iRule.From) > 0`?

joestringer replied:
Good catch, will fix.

aanm commented on pkg/maps/policymap/policymap.go:
```diff
@@ -137,7 +154,7 @@ func OpenMap(path string) (*PolicyMap, bool, error) {
 		path,
 		bpf.BPF_MAP_TYPE_HASH,
-		uint32(unsafe.Sizeof(uint32(0))),
+		uint32(unsafe.Sizeof(policyKey{})),
```
The key size is not the same, what are the consequences for running instances in case we upgrade?

joestringer replied:
Looks not great right now. I believe that after the new Cilium instance starts, it will attempt to regenerate the BPF (including PolicyMap) for existing endpoints, which will call down to OpenOrCreateMap(). If that latter function finds a map in the correct path, it won't validate that the map parameters are the same, so the existing map with a smaller valueSize will remain. Subsequently, when attempting to load the BPF program into the kernel, verification will fail due to access beyond the valueSize of the map. That will leave the container in a bad state.

I think that OpenOrCreateMap() should validate that all of the passed parameters are the same for the existing map, or delete it and attempt to create a new map to mitigate the above issue.

joestringer replied:
Looks like there's already some validation in an alternative, OpenOrCreate() but not all of the maps are using this. Should be a minor extra patch.

joestringer replied:
Commit https://github.com/cilium/cilium/pull/1599/commits/7c8f5024517cfb5a31f9251bd38e05a7513cf7c3 attempts to address this, would be keen to get some feedback on that.

aanm commented on pkg/k8s/network_policy_test.go:
```diff
@@ -23,6 +23,7 @@ import (
 	"github.com/cilium/cilium/pkg/policy"
 	"github.com/cilium/cilium/pkg/policy/api"
 
+	"fmt"
```
Move the `"fmt"` to line 18

joestringer replied:
done.

aanm commented on pkg/k8s/network_policy_v1beta_test.go:
```diff
@@ -24,6 +24,7 @@ import (
 	"github.com/cilium/cilium/pkg/policy"
 	"github.com/cilium/cilium/pkg/policy/api"
 
+	"fmt"
```
Move the "fmt" to line 20.

joestringer replied:
done.

aanm commented on pkg/k8s/network_policy_test.go:
```diff
@@ -95,12 +98,25 @@ func (s *K8sSuite) TestParseNetworkPolicy(c *C) {
 
+	matchLabels := make(map[string]string)
+	for _, v := range fromEndpoints {
+		matchLabels[fmt.Sprintf("%s.%s", v.Source, v.Key)] = v.Value
```
use the `GetExtendedKeyFrom`: `matchLabels[labels.GetExtendedKeyFrom(k)] = v.Value`

aanm commented on pkg/k8s/network_policy_v1beta_test.go:
```diff
@@ -87,12 +89,25 @@ func (s *K8sSuite) TestParseNetworkPolicyDeprecated(c *C) {
 
+	matchLabels := make(map[string]string)
+	for _, v := range fromEndpoints {
+		matchLabels[fmt.Sprintf("%s.%s", v.Source, v.Key)] = v.Value
```
use the `GetExtendedKeyFrom`: `matchLabels[labels.GetExtendedKeyFrom(k)] = v.Value`

aanm commented on pkg/policy/rule.go:
```diff
@@ -74,26 +75,70 @@ func (r *rule) validate() error {
+		return true
+	}
+
+	if len(policy.FromEndpoints) > 0 && len(fromEndpoints) == 0 {
```
What's the difference between this line and line 80?

joestringer replied:
Line 80 is saying that if the policy as resolved so far has an empty "FromEndpoints" list then it already covers the endpoints specified by the L4 filter that is currently being processed. Nil FromEndpoints would mean any endpoint can be used.

This line 85 is stating that if there are currently constraints on which endpoints this rule must receive traffic from in the policy, but the L4 Filter being processed allows contact from any Endpoint, then we need to loosen the current policy's restrictions on endpoints.

aanm commented on pkg/policy/api/selector.go:
```diff
@@ -20,6 +20,7 @@ import (
 
 	"github.com/cilium/cilium/pkg/labels"
 
+	"github.com/mitchellh/hashstructure"
```
Was this imported into `vendor/`?

joestringer replied:
I believe so, here's the commit which updates vendor.conf and imports the code:
https://github.com/cilium/cilium/pull/1599/commits/1ae6a2c2124c385c9257f006eaafc8ca414d5898

aanm commented on pkg/policy/repository.go:
```diff
@@ -147,6 +147,62 @@ func (p *Repository) ResolveL3Policy(ctx *SearchContext) *L3Policy {
 }
 
+func (p *Repository) allowsL4Egress(ctx *SearchContext) api.Decision {
+	ctx.To = ctx.From
```
Can you add a comment why you are doing this?

joestringer committed:
tests/cidr: Refactor timeouts

Put the timeouts in a convenient one-stop location to update to speed up
local testing. This also makes it easier to consider reducing these
counts in future to speed up CI runs.

This commit shortens the "drop" timeouts to 10s, as well. This is still
long compared to the CT tests' 2s DROP_TIMEOUT, but it's closer to the
current 14s.

Signed-off-by: Joe Stringer <joe@covalent.io>

joestringer committed:
tests/k8s: Flush L3/L4 policy before running L7

The GSG test was running the L7 portion of the test with both the L3/L4
policy, as well as the L3/L4/L7 policy, which could lead to unexpected
behaviour. Fix it.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on NEWS.rst:
```diff
@@ -15,6 +15,7 @@ Features
 * Add bash code completion (1597_)
 * L7 logging records now include additional information about the identity of
   the destination when outside of the cluster (1615_)
+* Support L3-dependent L4 policies on ingress (1599_, 1496_, 1217_, 1064_, 789_)
```
At the bottom of the NEWS.rst file, add the links that these PR's refer to.

joestringer replied:
Ah, I thought that was automagic. OK.

joestringer replied:
Ah, I thought that was automagic. OK.

tgraf replied:
I have a vim script to this automatically for links that are not complete yet.

ianvernon commented on pkg/policy/api/rule.go:
```diff
@@ -94,9 +94,18 @@ var EntitySelectorMapping = map[Entity]EndpointSelector{
-//   same IngressRule.
+//
+// - If multiple members are set, all of them need to match in order for
+//   the rule to take effect. The exception to this rule is FromRequires field,
```
... FromRequires field; the effects ...

joestringer replied:
Fixed

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -45,12 +58,14 @@ type L4Filter struct {
 	Port int
 	// Protocol is the L4 protocol to allow or NONE
 	Protocol string
+	// From Endpoints
```
incomplete documentation

joestringer replied:
Fixed

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -74,30 +75,78 @@ func (r *rule) validate() error {
+func adjustL4PolicyIfNeeded(fromEndpoints []api.EndpointSelector, policy *L4Filter) bool {
+
+	if len(policy.FromEndpoints) == 0 && len(fromEndpoints) > 0 {
+		log.Debugf("skipping this L4 policy since it is already covered by an existing policy.")
```
"this" is vague for a log message; which L4 Policy are we skipping? The log message should reflect which policy is being skipped.

joestringer replied:
I'll update it to print the policy and the fromEndpoints.

ianvernon commented on tests/k8s/helpers.bash:
```diff
@@ -5,6 +5,8 @@ source "${dir}/../helpers.bash"
 # dir might have been overwritten by helpers.bash
 dir=$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )
 
+MAX_LOG_PRINT=30
```
Thanks for adding this!

joestringer replied:
Ah, I'll actually drop this change out from this series because it's unrelated, but I do plan on following up on this separately.

ianvernon commented on tests/k8s/tests/03-l7-stresstest.sh:
```diff
@@ -1,168 +0,0 @@
-    log "k8s version is 1.6; checking that policies were deleted in Ciliumd"
-    docker exec -i ${cilium_id} cilium policy get io.cilium.k8s-policy-name=l7-stresstest-deprecated 2>/dev/null
-    if [ $? -eq 0 ]; then abort "l7-stresstest-deprecated policy found in cilium; policy should have been deleted" ; fi
-fi
```
Please bring back this file

joestringer replied:
Yeah, I'll drop the patch that deleted this when I push my latest changes.

joestringer replied:
This was an extra testing patch; I'll drop it.

joestringer committed:
daemon: Add map validation stage

Currently, the main daemon will typically attempt to rebuild the base
BPF programs before it performs any checks of the existing maps.
Unfortunately, that means that if we ever change attributes of a BPF
map, then we can easily break in-place upgrade:

- Existing Cilium is running with an old map layout
- Containers are currently attached with policy enforcement enabled
- New version of Cilium with different map layout is installed
- Cilium service is restarted
- On startup, Cilium calls into bpf/init.sh to compile the base programs
- init.sh calls tc lo load the tunnel BPF program with the new map layout
- The tc commandline validates the map info and sees that the layouts
  are different from the existing pinned map, and returns an error
- This error is propagated back up to the Cilium startup function
- Cilium reports the error and exits.

This patch attempts to address the above failure by performing a map
validation step during daemon initialization, before the call to rebuild
the base BPF programs. It iterates through the maps in the filesystem
and if any Cilium PolicyMaps are pinned with attributes that differ from
the current attributes, it deletes these maps.

Signed-off-by: Joe Stringer <joe@covalent.io>

ianvernon commented on Documentation/policy.rst:
```diff
@@ -669,10 +669,10 @@ endpoint with the label ``id.http`` on port 80:
         Allows from labels {"matchLabels":{"any:id.curl":""}}
-    +     Found all required labels
+          Found all required labels
+            Restricted to specific L4 destinations; skip during Label stage
```
Along with this change to the tracing output, I think we should add documentation to this file explaining L3-dependent L4 policy as well. I don't mind this being done after this patch is merged in the interest of having you not have to rebase again :)

joestringer replied:
I'll take note to follow up on this. I think that the way that the documentation is written right now, it already implies the behaviour specified with this l3-dependent l4 work. I just don't think that we have any examples using multiple ingress rules, so we could highlight the behaviour difference. This could help people who end up in a situation where they're allowing L3 and allowing L4, but actually intended to only specify l3-dependent l4 policy.

ianvernon replied:
Great, sounds good! Please create a doc ticket for that so we have it tracked.

joestringer replied:
Done: https://github.com/cilium/cilium/issues/1685

ianvernon commented on pkg/policy/api/rule.go:
```diff
@@ -94,9 +94,18 @@ var EntitySelectorMapping = map[Entity]EndpointSelector{
+//   the effects of any Requires field in any rule will apply to all other
+//   rules as well.
+//
+// - For now, combining ToPorts, FromCIDR, and FromEndpoints in the same rule
```
Is there a GitHub issue for this 'in the future' task? Please create one if it does not exist.

joestringer replied:
Done: https://github.com/cilium/cilium/issues/1684

tgraf commented on daemon/daemon.go:
```diff
@@ -1027,6 +1045,30 @@ func (d *Daemon) staleMapWalker(path string) error {
 	return nil
 }
 
+func (d *Daemon) mapValidateWalker(path string) error {
```
I don't think you need to make this a function of `Daemon`. Let's just keep it private. Longer term I want to get rid of the `Daemon struct`

tgraf replied:
Can be done as follow-up

ianvernon commented on pkg/policy/rule.go:
```diff
@@ -94,32 +112,63 @@ func mergeL4Port(ctx *SearchContext, r api.PortRule, p api.PortProtocol, dir str
+	for hash, newL7Rules := range l4Filter.L7RulesPerEp {
+		if ep, ok := v.L7RulesPerEp[hash]; ok {
+			switch {
+			case len(l4Filter.L7RulesPerEp[hash].HTTP) > 0:
```
This is just something to note - how will we handle these types of 'collisions' in the future as more and more protocols are supported?

joestringer replied:
Definitely something we should look at addressing by the time we consider supporting a third protocol.

ianvernon commented on tests/16-cidr-ingress-policy.sh:
```diff
@@ -68,7 +71,7 @@ monitor_clear
 
 set +e
-docker run --rm -i --net ${TEST_NET} -l "${ID_SERVICE2}" --cap-add NET_ADMIN ${DEMO_CONTAINER} ping -c 14 ${IPV4_HOST} && {
+docker run --rm -i --net ${TEST_NET} -l "${ID_SERVICE2}" --cap-add NET_ADMIN ${DEMO_CONTAINER} ping -c ${DROP_TIMEOUT} ${IPV4_HOST} && {
```
+1 for factoring stuff out into constants.

ianvernon reviewed (approved):
This looks *fantastic*. Testing coverage looks good, and the implementation is elegant. Great job! I'm approving, but please make sure that those who have also reviewed approve before merging :)
